lucene/analysis/common/src/java/org/apache/lucene/analysis/pattern/PatternTypingFilter.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.lucene.analysis.pattern;
+
+import org.apache.lucene.analysis.TokenFilter;
+import org.apache.lucene.analysis.TokenStream;
+import org.apache.lucene.analysis.tokenattributes.CharTermAttribute;
+import org.apache.lucene.analysis.tokenattributes.FlagsAttribute;
+import org.apache.lucene.analysis.tokenattributes.TypeAttribute;
+
+import java.io.IOException;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * Set a type attribute to a parameterized value when tokens are matched by any of a several regex patterns. The
+ * value set in the type attribute is parameterized with the match groups of the regex used for matching.
+ * In combination with TypeAsSynonymFilter and DropIfFlagged filter this can supply complex synonym patterns
+ * that are protected from subsequent analysis, and optionally drop the original term based on the flag
+ * set in this filter. See {@link PatternTypingFilterFactory} for full documentation.
+ *
+ * @see PatternTypingFilterFactory
+ * @since 8.8.0
+ */
+public class PatternTypingFilter extends TokenFilter {
+
+  private final Map<Pattern, Map.Entry<String, Integer>> replacementAndFlagByPattern;
+  private final CharTermAttribute termAtt = addAttribute(CharTermAttribute.class);
+  private final FlagsAttribute flagAtt = addAttribute(FlagsAttribute.class);
+  private final TypeAttribute typeAtt = addAttribute(TypeAttribute.class);
+
+  public PatternTypingFilter(TokenStream input, LinkedHashMap<Pattern, Map.Entry<String, Integer>> replacementAndFlagByPattern) {
```

Comment:
Explicitely saying LinkedHashMap sounds strange. I know the list must be sorted, so it's more a list.

I liked it use a <Pattern,String,Integer> record. This would be a classical exple of the new Java 6 record types!

This is a public API, any client code may call this - also non-Solr users. So maybe the constructor argument should be `a List<PatternTypingRule>` (a new class, which may actually be a Record in Java 16/17).

Reply:
FYI, I was not aware that the Map of Patterns and Elements are exposed as public API.

Reply:
If we have a class `PatternTypingRule` we can make constructors to create those: `new PatternTypeingRule(Pattern pattern, String replacement, int flag)` but also `new PatternTypingRule(String pattern, String replacement, int flag)`.

I would strogly prefer tono misuse maps

Reply:
Updated. I agree that this is cleaner. I've started getting some time pressure on this due to end of month approaching so hopefully if you like what I've done we can get this committed. I think we've made good improvements and further refinements have time to be contemplated before 8.8. Getting it in will allow an updated patch for the primary AQP feature.